### ISSUE TITLE
NoClassDefFound after updating gcloud to 159.0.0

### ISSUE THREAD

**Issue by Radagan** [GoogleCloudPlatform/google-cloud-intellij#1533]
I updated gcloud to 159.0.0 and deployed a version of my app engine app. To be clear there are no source changes and no source uploads. Once deployed none of the classes can be found. When I roll back using `gcloud components update --version 158.0.0` everything works fine again. 

I'm running IDEA 2017.1.4 and Google Cloud Tools plugin 17.6.2.

Any idea what might cause this? There are no errors when I update to 159, and rolling back fixes the issue every time, but I'm stuck at 158. I did try updated to 161 today, but it has the same issue.

**Comment by elharo** [comment]
Can you provide the actual error message, screenshot, and stack trace?

**Comment by Radagan** [comment]
That's just it, there isn't any that I can find. Here's the console output from the plugin:

> Deploying '[2017-07-05 11:27:38] BuildingPulse:war exploded. Project: buildingpulse. Version: auto'...
> Created temporary staging directory: /private/var/folders/ws/tqjggzd938g4trz89djn_x8w0000gn/T/gae-staging-buildingpulse2
> Reading application configuration data...
> Jul 05, 2017 11:27:40 AM com.google.apphosting.utils.config.IndexesXmlReader readConfigXml
> INFO: Successfully processed /Users/Radagan/Development/Java/BuildingPulse/out/artifacts/BuildingPulse_war_exploded/WEB-INF/datastore-indexes.xml
> 
> 
> Beginning interaction for module default...
> 0% Scanning for jsp files.
> 0% Compiling jsp files.
> Jul 05, 2017 11:27:41 AM org.apache.jasper.JspC processFile
> INFO: Built File: /index.jsp
> 0% Generated git repository information file.
> Jul 05, 2017 11:27:41 AM com.google.appengine.tools.admin.Application createStagingDirectory
> Success.
> INFO: We detected that you have a large number of .class files in WEB-INF/classes. You may be able to reduce request latency by packaging your .class files into jars. To do this, supply the --enable_jar_classes flag when using appcfg on command line. If you're using the gcloud-maven-plugin, add <enable_jar_classes>true</enable_jar_classes> in the plugin's <configuration> tag. If you are using the appengine-maven-plugin, supply <enableJarClasses>true</enableJarClasses> in the plugin's <configuration> tag. Note that this flag will put the jar in WEB-INF/lib rather than WEB-INF/classes. The classloader first looks in WEB-INF/classes and then WEB-INF/lib when loading a class. As a result, this flag could change classloading order, which may affect the behavior of your app.
> Temporary staging for module default directory left in /private/var/folders/ws/tqjggzd938g4trz89djn_x8w0000gn/T/gae-staging-buildingpulse2
> You are about to deploy the following services:
>  - buildingpulse/default/20170705t112744 (from [/private/var/folders/ws/tqjggzd938g4trz89djn_x8w0000gn/T/gae-staging-buildingpulse2/app.yaml])
>      Deploying to URL: [https://buildingpulse.appspot.com]
> 
> You are about to update the following configurations:
>  - buildingpulse/cron  (from [/private/var/folders/ws/tqjggzd938g4trz89djn_x8w0000gn/T/gae-staging-buildingpulse2/WEB-INF/appengine-generated/cron.yaml])
>  - buildingpulse/index  (from [/private/var/folders/ws/tqjggzd938g4trz89djn_x8w0000gn/T/gae-staging-buildingpulse2/WEB-INF/appengine-generated/index.yaml])
>  - buildingpulse/queue  (from [/private/var/folders/ws/tqjggzd938g4trz89djn_x8w0000gn/T/gae-staging-buildingpulse2/WEB-INF/appengine-generated/queue.yaml])
> Beginning deployment of service [default]...
> Some files were skipped. Pass `--verbosity=info` to see which ones.
> You may also view the gcloud log file, found at
> [/Users/Radagan/.config/gcloud/logs/2017.07.05/11.27.42.024609.log].
> #============================================================#
> #= Uploading 0 files to Google Cloud Storage                =#
> #============================================================#
> File upload done.
> Updating service [default]...
> .....................done.
> Deployed service [default] to [https://buildingpulse.appspot.com]
> Updating config [cron]...
> .done.
> Updating config [index]...
> .done.
> Updating config [queue]...
> .done.
> 
> Cron jobs have been updated.
> 
> Task queues have been updated.
> 
> Visit the Cloud Platform Console Task Queues page to view your queues and cron jobs.
> 
> Indexes are being rebuilt. This may take a moment.
> 
> You can stream logs from the command line by running:
>   $ gcloud app logs tail -s default
> 
> To view your application in the web browser run:
>   $ gcloud app browse
> '[2017-07-05 11:27:38] BuildingPulse:war exploded. Project: buildingpulse. Version: auto' has been deployed successfully.

When I look at the logs for the version in Chrome, any use of my project's classes throws a NoClassDefFoundError. So even though the new version does not update those classes, they can't be found at runtime. 

All I have to do to reproduce a working version is roll back to 158.0.0 in gcloud and hit deploy in Idea again. Upgrade to any version above 158.0.0 and I get the behavior above.

Here's a stack trace from the logs showing the error:

> 
> java.lang.NoClassDefFoundError: com/google/appengine/api/taskqueue/DeferredTask
> 	at net.logicalbuilding.buildingpulse.AccessServlet.doWork(AccessServlet.java:48)
> 	at net.logicalbuilding.buildingpulse.BaseServlet.doPost(BaseServlet.java:84)
> 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:637)
> 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
> 	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
> 	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
> 	at com.google.apphosting.utils.servlet.ParseBlobUploadFilter.doFilter(ParseBlobUploadFilter.java:125)
> 	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
> 	at com.google.apphosting.runtime.jetty.SaveSessionFilter.doFilter(SaveSessionFilter.java:37)
> 	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
> 	at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
> 	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
> 	at com.google.apphosting.utils.servlet.TransactionCleanupFilter.doFilter(TransactionCleanupFilter.java:48)
> 	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
> 	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
> 	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
> 	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
> 	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
> 	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
> 	at com.google.apphosting.runtime.jetty.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:257)
> 	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
> 	at org.mortbay.jetty.Server.handle(Server.java:326)
> 	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
> 	at org.mortbay.jetty.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:923)
> 	at com.google.apphosting.runtime.jetty.RpcRequestParser.parseAvailable(RpcRequestParser.java:76)
> 	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
> 	at com.google.apphosting.runtime.jetty.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:146)
> 	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.dispatchServletRequest(JavaRuntime.java:657)
> 	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.dispatchRequest(JavaRuntime.java:619)
> 	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:589)
> 	at com.google.tracing.TraceContext$TraceContextRunnable.runInContext(TraceContext.java:453)
> 	at com.google.tracing.TraceContext$TraceContextRunnable$1.run(TraceContext.java:460)
> 	at com.google.tracing.CurrentContext.runInContext(CurrentContext.java:293)
> 	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContextNoUnref(TraceContext.java:319)
> 	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContext(TraceContext.java:311)
> 	at com.google.tracing.TraceContext$TraceContextRunnable.run(TraceContext.java:457)
> 	at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:263)
>

**Comment by etanshaul** [comment]
@Radagan do you happen to have (or could quickly prepare) a trimmed down example app that could reproduce this error? That would make it easier for us to debug. If you could post it on github that would be very helpful. I know its not ideal, but there's not much to go off of from those errors.

**Comment by Radagan** [comment]
Not easily because the datastore seems to be involved. Can you think if any differences in dependancies in gcloud between 158 and 159? It seems like it has to be related to the gcloud foundation your calling from the your plugin. Maybe I don't have something installed that I now need in 159?

Here's what I have:

> Your current Cloud SDK version is: 159.0.0
> The latest available version is: 161.0.0
> 
> ┌────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
> │                                                   Components                                                   │
> ├──────────────────┬──────────────────────────────────────────────────────┬──────────────────────────┬───────────┤
> │      Status      │                         Name                         │            ID            │    Size   │
> ├──────────────────┼──────────────────────────────────────────────────────┼──────────────────────────┼───────────┤
> │ Update Available │ Cloud SDK Core Libraries                             │ core                     │   6.4 MiB │
> │ Update Available │ gcloud app Java Extensions                           │ app-engine-java          │ 128.1 MiB │
> │ Update Available │ gcloud app Python Extensions                         │ app-engine-python        │   6.5 MiB │
> │ Not Installed    │ Cloud Datastore Emulator (Legacy)                    │ gcd-emulator             │  38.1 MiB │
> │ Installed        │ App Engine Go Extensions                             │ app-engine-go            │  96.7 MiB │
> │ Installed        │ BigQuery Command Line Tool                           │ bq                       │   < 1 MiB │
> │ Installed        │ Cloud Bigtable Command Line Tool                     │ cbt                      │   4.0 MiB │
> │ Installed        │ Cloud Bigtable Emulator                              │ bigtable                 │   3.5 MiB │
> │ Installed        │ Cloud Datalab Command Line Tool                      │ datalab                  │   < 1 MiB │
> │ Installed        │ Cloud Datastore Emulator                             │ cloud-datastore-emulator │  15.4 MiB │
> │ Installed        │ Cloud Pub/Sub Emulator                               │ pubsub-emulator          │  21.0 MiB │
> │ Installed        │ Cloud Storage Command Line Tool                      │ gsutil                   │   2.9 MiB │
> │ Installed        │ Default set of gcloud commands                       │ gcloud                   │           │
> │ Installed        │ Emulator Reverse Proxy                               │ emulator-reverse-proxy   │  14.5 MiB │
> │ Installed        │ Google Container Registry's Docker credential helper │ docker-credential-gcr    │   2.2 MiB │
> │ Installed        │ gcloud Alpha Commands                                │ alpha                    │   < 1 MiB │
> │ Installed        │ gcloud Beta Commands                                 │ beta                     │   < 1 MiB │
> │ Installed        │ gcloud app PHP Extensions (Mac OS X)                 │ app-engine-php-darwin    │  21.9 MiB │
> │ Installed        │ kubectl                                              │ kubectl                  │  14.8 MiB │
> └──────────────────┴──────────────────────────────────────────────────────┴──────────────────────────┴───────────┘
> To install or remove components at your current SDK version [159.0.0], run:
>   $ gcloud components install COMPONENT_ID
>   $ gcloud components remove COMPONENT_ID
> 
> To update your SDK installation to the latest version [161.0.0], run:
>   $ gcloud components update
> 
>

**Comment by etanshaul** [comment]
Can you run `gcloud --version` using a failing version of gcloud from the command line and paste the output here?

e.g.:
![image](https://user-images.githubusercontent.com/1735744/27919496-cfce5932-623f-11e7-8043-696de2d289c7.png)

I'm interested in the version of app-engine-java you are using.

**Comment by etanshaul** [comment]
Could you also please paste the contents of your `appengine-web.xml` file (removing any identifying info from it)?

**Comment by Radagan** [comment]
And to think I was foolish enough to trust Google with an update task... You put your finger on it: gcloud installs the wrong version of app-engine-java until the most recent 161. 

```
You will be upgraded to version: 159.0.0

┌───────────────────────────────────────────────────────────────────────┐
│                   These components will be updated.                   │
├──────────────────────────────────────────────┬────────────┬───────────┤
│                     Name                     │  Version   │    Size   │
├──────────────────────────────────────────────┼────────────┼───────────┤
│ Cloud SDK Core Libraries                     │ 2017.06.09 │   6.2 MiB │
│ Cloud SDK Core Libraries (Platform Specific) │ 2017.05.30 │   < 1 MiB │
│ cbt (Mac OS X, x86_64)                       │      0.2.1 │   4.0 MiB │
│ gcloud app Java Extensions                   │     1.9.53 │ 132.2 MiB │
│ gcloud app Python Extensions                 │     1.9.54 │   6.4 MiB │
│ gcloud cli dependencies                      │ 2017.06.09 │   1.6 MiB │
│ kubectl (Mac OS X, x86_64)                   │      1.6.4 │  14.8 MiB │
└──────────────────────────────────────────────┴────────────┴───────────┘

For the latest full release notes, please visit:
  https://cloud.google.com/sdk/release_notes

Do you want to continue (Y/n)?  n

Athyra:Java Radagan$ ./google-cloud-sdk/bin/gcloud components update --version 160.0.0


Your current Cloud SDK version is: 161.0.0
You will be upgraded to version: 160.0.0

┌───────────────────────────────────────────────────────┐
│           These components will be updated.           │
├──────────────────────────────┬────────────┬───────────┤
│             Name             │  Version   │    Size   │
├──────────────────────────────┼────────────┼───────────┤
│ Cloud SDK Core Libraries     │ 2017.06.16 │   6.3 MiB │
│ gcloud app Java Extensions   │     1.9.53 │ 132.2 MiB │
│ gcloud app Python Extensions │     1.9.54 │   6.4 MiB │
│ gcloud cli dependencies      │ 2017.06.09 │   1.6 MiB │
│ kubectl (Mac OS X, x86_64)   │      1.6.4 │  14.8 MiB │
└──────────────────────────────┴────────────┴───────────┘

For the latest full release notes, please visit:
  https://cloud.google.com/sdk/release_notes

Do you want to continue (Y/n)?  n   

Athyra:Java Radagan$ ./google-cloud-sdk/bin/gcloud components update --version 161.0.0

All components are up to date.
Athyra:Java Radagan$ ./google-cloud-sdk/bin/gcloud --version
Google Cloud SDK 161.0.0
alpha 2017.03.24
app-engine-go 
app-engine-java 1.9.54
app-engine-php-darwin 2017.03.24
app-engine-python 1.9.55
beta 2017.03.24
bigtable 
bq 2.0.24
cbt 
cloud-datastore-emulator 1.2.1
core 2017.06.26
datalab 20170525
docker-credential-gcr 
emulator-reverse-proxy 
gcloud 
gsutil 4.26
kubectl 
pubsub-emulator 2017.03.24
Athyra:Java Radagan$ 
```

I wonder if Amazon is this buggy and procedurally chaotic. Now for some reason my build is failing for a different reason: Upload error 503. Looking into it...

**Comment by Radagan** [comment]
The 503 was nothing--I just needed to re-auth. Now for the best part: **nothing has changed**. 

I get the exact same failure with 161.0.0 and it is using the correct app-engine-java... Below are the files you requested. Thank you again for helping me with this. It's a major problem for us on an actual commercial product in the building automation market. The app has been running with continual care and feeding for almost four years. 

```
Athyra:Java Radagan$ ./google-cloud-sdk/bin/gcloud --version
Google Cloud SDK 161.0.0
alpha 2017.03.24
app-engine-go 
app-engine-java 1.9.54
app-engine-php-darwin 2017.03.24
app-engine-python 1.9.55
beta 2017.03.24
bigtable 
bq 2.0.24
cbt 
cloud-datastore-emulator 1.2.1
core 2017.06.26
datalab 20170525
docker-credential-gcr 
emulator-reverse-proxy 
gcloud 
gsutil 4.26
kubectl 
pubsub-emulator 2017.03.24
```

appengine-web.xml:
```
<?xml version="1.0" encoding="utf-8"?>
<appengine-web-app xmlns="http://appengine.google.com/ns/1.0">
  <!--
    Allows App Engine to send multiple requests to one instance in parallel:
  -->
  <threadsafe>true</threadsafe>

  <!-- Configure java.util.logging -->
  <system-properties>
    <property name="java.util.logging.config.file" value="WEB-INF/logging.properties"/>
  </system-properties>

  <!--
    HTTP Sessions are disabled by default. To enable HTTP sessions specify:

      <sessions-enabled>true</sessions-enabled>

    It's possible to reduce request latency by configuring your application to
    asynchronously write HTTP session data to the datastore:

      <async-session-persistence enabled="true" />

    With this feature enabled, there is a very small chance your app will see
    stale session data. For details, see
    http://code.google.com/appengine/docs/java/config/appconfig.html#Enabling_Sessions
  -->

</appengine-web-app>

```

Is there some way to force gcloud to upload a clean build? 

```
Deploying '[2017-07-07 11:26:30] BuildingPulse:war exploded. Project: buildingpulse. Version: auto'...
Created temporary staging directory: /private/var/folders/ws/tqjggzd938g4trz89djn_x8w0000gn/T/gae-staging-buildingpulse
Reading application configuration data...
Jul 07, 2017 11:26:32 AM com.google.apphosting.utils.config.IndexesXmlReader readConfigXml
INFO: Successfully processed /Users/Radagan/Development/Java/BuildingPulse/out/artifacts/BuildingPulse_war_exploded/WEB-INF/datastore-indexes.xml


Beginning interaction for module default...
0% Scanning for jsp files.
0% Compiling jsp files.
Jul 07, 2017 11:26:32 AM org.apache.jasper.JspC processFile
INFO: Built File: /index.jsp
0% Generated git repository information file.
Success.
Jul 07, 2017 11:26:33 AM com.google.appengine.tools.admin.Application createStagingDirectory
Temporary staging for module default directory left in /private/var/folders/ws/tqjggzd938g4trz89djn_x8w0000gn/T/gae-staging-buildingpulse
INFO: We detected that you have a large number of .class files in WEB-INF/classes. You may be able to reduce request latency by packaging your .class files into jars. To do this, supply the --enable_jar_classes flag when using appcfg on command line. If you're using the Cloud SDK based app-maven-plugin, add <stage.enableJarClasses>true</stage.enableJarClasses> in the plugin's <configuration> tag. If you are using the AppCfg based appengine-maven-plugin, supply <enableJarClasses>true</enableJarClasses> in the plugin's <configuration> tag. Note that this flag will put the jar in WEB-INF/lib rather than WEB-INF/classes. The classloader first looks in WEB-INF/classes and then WEB-INF/lib when loading a class. As a result, this flag could change classloading order, which may affect the behavior of your app.
Services to deploy:

descriptor:      [/private/var/folders/ws/tqjggzd938g4trz89djn_x8w0000gn/T/gae-staging-buildingpulse/app.yaml]
source:          [/private/var/folders/ws/tqjggzd938g4trz89djn_x8w0000gn/T/gae-staging-buildingpulse]
target project:  [buildingpulse]
target service:  [default]
target version:  [20170707t112636]
target url:      [https://buildingpulse.appspot.com]


Configurations to update:

descriptor:      [/private/var/folders/ws/tqjggzd938g4trz89djn_x8w0000gn/T/gae-staging-buildingpulse/WEB-INF/appengine-generated/index.yaml]
type:            [datastore indexes]
target project:  [buildingpulse]


descriptor:      [/private/var/folders/ws/tqjggzd938g4trz89djn_x8w0000gn/T/gae-staging-buildingpulse/WEB-INF/appengine-generated/cron.yaml]
type:            [cron jobs]
target project:  [buildingpulse]


descriptor:      [/private/var/folders/ws/tqjggzd938g4trz89djn_x8w0000gn/T/gae-staging-buildingpulse/WEB-INF/appengine-generated/queue.yaml]
type:            [task queues]
target project:  [buildingpulse]


Beginning deployment of service [default]...
Some files were skipped. Pass `--verbosity=info` to see which ones.
You may also view the gcloud log file, found at
[/Users/Radagan/.config/gcloud/logs/2017.07.07/11.26.33.816740.log].
#============================================================#
#= Uploading 0 files to Google Cloud Storage                =#
#============================================================#
File upload done.
Updating service [default]...
...................done.
Deployed service [default] to [https://buildingpulse.appspot.com]
Updating config [index]...
.done.
Updating config [cron]...
.done.
Updating config [queue]...
.done.

Cron jobs have been updated.

Task queues have been updated.

Visit the Cloud Platform Console Task Queues page to view your queues and cron jobs.

Indexes are being rebuilt. This may take a moment.

You can stream logs from the command line by running:
  $ gcloud app logs tail -s default

To view your application in the web browser run:
  $ gcloud app browse
'[2017-07-07 11:26:30] BuildingPulse:war exploded. Project: buildingpulse. Version: auto' has been deployed successfully.
```

**Comment by Radagan** [comment]
Just for completeness, when the new version is promoted, nothing but NoClassDefFoundErrors in the logs like this one:

```
Error for /access
java.lang.NoClassDefFoundError: com/google/appengine/api/taskqueue/DeferredTask
	at net.logicalbuilding.buildingpulse.AccessServlet.doWork(AccessServlet.java:48)
	at net.logicalbuilding.buildingpulse.BaseServlet.doPost(BaseServlet.java:84)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:637)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at com.google.apphosting.utils.servlet.ParseBlobUploadFilter.doFilter(ParseBlobUploadFilter.java:125)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.runtime.jetty.SaveSessionFilter.doFilter(SaveSessionFilter.java:37)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.TransactionCleanupFilter.doFilter(TransactionCleanupFilter.java:48)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at com.google.apphosting.runtime.jetty.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:257)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:923)
	at com.google.apphosting.runtime.jetty.RpcRequestParser.parseAvailable(RpcRequestParser.java:76)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at com.google.apphosting.runtime.jetty.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:146)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.dispatchServletRequest(JavaRuntime.java:657)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.dispatchRequest(JavaRuntime.java:619)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:589)
	at com.google.tracing.TraceContext$TraceContextRunnable.runInContext(TraceContext.java:453)
	at com.google.tracing.TraceContext$TraceContextRunnable$1.run(TraceContext.java:460)
	at com.google.tracing.CurrentContext.runInContext(CurrentContext.java:293)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContextNoUnref(TraceContext.java:319)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContext(TraceContext.java:311)
	at com.google.tracing.TraceContext$TraceContextRunnable.run(TraceContext.java:457)
	at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:263)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ClassNotFoundException: com.google.appengine.api.taskqueue.DeferredTask
	at com.google.appengine.runtime.Request.process-62364c5f057fdb17(Request.java)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:375)
	at net.logicalbuilding.buildingpulse.AccessServlet.doWork(AccessServlet.java:48)
	at net.logicalbuilding.buildingpulse.BaseServlet.doPost(BaseServlet.java:84)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:637)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:923)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at com.google.tracing.TraceContext$TraceContextRunnable.runInContext(TraceContext.java:453)
	at com.google.tracing.TraceContext$TraceContextRunnable$1.run(TraceContext.java:460)
	at com.google.tracing.CurrentContext.runInContext(CurrentContext.java:293)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContextNoUnref(TraceContext.java:319)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContext(TraceContext.java:311)
	at com.google.tracing.TraceContext$TraceContextRunnable.run(TraceContext.java:457)

```

**Comment by elharo** [comment]
Is it always com/google/appengine/api/taskqueue/DeferredTask that is not found, or other classes in com/google/appengine/api? That sounds like the appengine-api-1.0-sdk.jar is not getting added to the classpath.

**Comment by etanshaul** [comment]
@radagan this is being looked at internally by the appropriate team

**Comment by Radagan** [comment]
I’m almost certain that it’s all classes. That one is just common because it results from a frequent CRON that maintains status with a number of sites (actual office buildings). I’ll make sure though, and update the issue with an example first thing on Monday morning.

**Comment by betamos** [comment]
Radagan, would you care to provide the gcloud log files from both the failing and the succeeding invocations? You can find them by inspecting the log message from your invocations, e.g. from above:

```
You may also view the gcloud log file, found at
[/Users/Radagan/.config/gcloud/logs/2017.07.07/11.26.33.816740.log].
```

Even better would be a (stripped down) version of an app that fails like this, but we can start with the logs.

**Comment by ludoch** [comment]
One thing I would like to see is the generated app.yaml at the root of the
deployment staging directory.
If it does not contain

api_version: '1.0'

it could be a bug in the IDE plugin integration.

**Comment by ludoch** [comment]
internal reference: 63439056

**Comment by Radagan** [comment]
@betamos, I noticed a new release of gcloud this morning, so I've done a clean rollback and then update for you. Unfortunately, 162 is no better in my case. Hopefully, these logs will help you. Thank you and all the others here who are looking into this problem. 

Step by step:

```
Getting back to a functional baseline:

$ ./google-cloud-sdk/bin/gcloud components update --version 158.0.0
/Users/Radagan/.config/gcloud/logs/2017.07.13/09.57.01.449031.log

This triggered an actual upload as it should: (from plugin in Idea’s console)
#============================================================#
#= Uploading 239 files to Google Cloud Storage              =#
#============================================================#
File upload done.

Verified that App runs as expected. No problems.

Now I update to current. There is a new one, so maybe this will work...

$ ./google-cloud-sdk/bin/gcloud components update

Done.

Now for a quick version check:

Athyra:Java Radagan$ ./google-cloud-sdk/bin/gcloud --version
Google Cloud SDK 162.0.0
alpha 2017.03.24
app-engine-go 
app-engine-java 1.9.54
app-engine-php-darwin 2017.03.24
app-engine-python 1.9.57
beta 2017.03.24
bigtable 
bq 2.0.24
cbt 
cloud-datastore-emulator 1.2.1
core 2017.07.07
datalab 20170525
docker-credential-gcr 
emulator-reverse-proxy 
gcloud 
gsutil 4.27
kubectl 
pubsub-emulator 2017.03.24

In Idea, I run a rebuild and then deploy

/Users/Radagan/.config/gcloud/logs/2017.07.13/10.07.28.780439.log

Nothing actually uploads though it is a success…
#============================================================#
#= Uploading 0 files to Google Cloud Storage                =#
#============================================================#
File upload done.

Now no one can even login, so I quickly transition back to the previous version in the App Engine configuration in Chrome for my project.
```

Here are all the logs, including those generated by the updates and other gcloud commands issued. I thought those might be useful too. See above for the specific upload logs. 

[2017.07.13.zip](https://github.com/GoogleCloudPlatform/google-cloud-intellij/files/1143706/2017.07.13.zip)

**Comment by Radagan** [comment]
@ludoch, no problem. Here is the app.yaml from the failing build above:

```
/private/var/folders/ws/tqjggzd938g4trz89djn_x8w0000gn/T/gae-staging-buildingpulse1/
```
[app.yaml.zip](https://github.com/GoogleCloudPlatform/google-cloud-intellij/files/1143720/app.yaml.zip)

I had to zip it to upload it here. It does not contain api_version: '1.0', but api_version: 'none'.

**Comment by chanseokoh** [comment]
> This triggered an actual upload as it should: (from plugin in Idea’s console)
> 
> ```
> #============================================================#
> #= Uploading 239 files to Google Cloud Storage              =#
> #============================================================#
> File upload done.
> ```
> 
> Nothing actually uploads though it is a success…
> ```
> #============================================================#
> #= Uploading 0 files to Google Cloud Storage                =#
> #============================================================#
> File upload done.
> ```

FYI, this is working as expected. The deploy process is smart enough to know that there is nothing to re-upload because your app hasn't changed.

If you do want to re-upload everything, just empty the Google Cloud Storage bucket where these files are being uploaded. If you haven't changed anything about the bucket, the bucket name is `staging.<your project name>.appspot.com`. You'll be able to see it in the Google Cloud Console. Files in the bucket will have machine-generated alphanumeric names like `4bd24f3fa461ce4f07160c334fc69d1686559098`. If you empty the bucket and deploy your app, I expect the bucket will have 239 files.

**Comment by chanseokoh** [comment]
About the staging bucket: https://cloud.google.com/appengine/docs/standard/go/googlecloudstorageclient/setting-up-cloud-storage

**Comment by Radagan** [comment]
@chanseokoh, Thank you for the info about the staging bucket. I've been wondering for a while how to do real, honest, **clean** build to app engine! 

Btw, I understand what your saying about no files uploading, I just wanted to highlight the point that nothing in my code was changing here. The only thing that command needed to cause complete runtime failure of my app, with no errors from Idea or gcloud, is to run:

```
$ ./google-cloud-sdk/bin/gcloud components update
```

... and hit Google App Engine Deploy in Idea.

**Comment by ludoch** [comment]
Assuming you have the correct AppEngine API jar in your WEB-INF/lib directory (before staging), the generated app.yaml should contain 1.0 instead of none....
One more request: can you share the ls -l of your WEB-INF/lib directory?
And even better, the content of the manifest in the appengine API jar that you should have there? The logic depends on the manifest content...

**Comment by chanseokoh** [comment]
I am also curious if the `app.yaml` generated when using gcloud 158 is identical to that by >158.

**Comment by Radagan** [comment]
Found it! 

For those like me who like details... When I did my clean update to 162, it should have worked. It now works perfectly on 162. The real problem was that after 158, for whatever reason app_engine 53 was being used instead of 54. It wasn't until I removed the reliance on maven Releases, unselected app engine in project settings for module "Google App Engine", and explicitly specified the maven 54 library version that gcloud showed the correct app engine in gcloud --version. 

But guess what I forgot to do... Yup: go back and select app engine in project settings. The yaml is correct now, and the build runs smoothly.

While I can't say I'm looking forward to what happens when app_engine 55 comes out, this experience has taught me quite a bit about how to troubleshoot these builds. Thank you very much for your time and patience.
